src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java

```diff
@@ -462,15 +462,34 @@ private static boolean isNonLocalTypeDeclaration(DetailAST typeDeclAst) {
     private static DetailAST getBlockContainingLocalAnonInnerClass(DetailAST literalNewAst) {
         DetailAST currentAst = literalNewAst;
         DetailAST result = null;
-        while (!TokenUtil.isOfType(currentAst, CONTAINERS_FOR_ANON_INNERS)) {
-            if (currentAst.getType() == TokenTypes.LAMBDA
-                    && currentAst.getParent()
-                    .getParent().getParent().getType() == TokenTypes.OBJBLOCK) {
+        while (result == null) {
+            if (!isTokenNestedUnderContainerForAnonInners(currentAst)) {
                 result = currentAst;
-                break;
             }
             currentAst = currentAst.getParent();
-            result = currentAst;
+        }
+        return result;
+    }
+
+    /**
+     * Whether the ast is nested under tokens of type
+     * {@link UnusedLocalVariableCheck#CONTAINERS_FOR_ANON_INNERS} or {@link TokenTypes#LAMBDA}.
+     *
+     * @param ast ast node
+     * @return {@code true} if ast is nested under tokens of type
+     *         {@link UnusedLocalVariableCheck#CONTAINERS_FOR_ANON_INNERS}
+     *         or {@link TokenTypes#LAMBDA}
+     */
+    public static boolean isTokenNestedUnderContainerForAnonInners(DetailAST ast) {
+        boolean result = false;
+        DetailAST parentAst = ast.getParent();
+        while (parentAst != null) {
+            if (TokenUtil.isOfType(parentAst, CONTAINERS_FOR_ANON_INNERS)
```

Comment:
Please rename this collection to something like `CLASS_BODY_DECLARATION_TOKENS` in a separate commit first, `container` is nonsense in this context. Then update the method naming, etc. appropriately.

Reply:
@rdiachenko , please step in to help review.

Reply:
Created https://github.com/checkstyle/checkstyle/pull/14090 to change the variable name.